### ISSUE TITLE
Creating a splitter with dock panels.

### ISSUE THREAD

**Issue by carlosdr02** [ocornut/imgui#5998]
Hi, I've been looking into creating a vertical splitter inside a window, I tried `SplitterBehaviour()`, but that seems to be only visual with no functionality, so now I'm trying to do it with dock panels, this is what I have now:

![image](https://user-images.githubusercontent.com/80007185/207661249-315e6309-2dcf-4447-82e5-3c2c21d53207.png)
I don't know if docking is the right way to do splitters, but is the only thing I can come up with.

But obviously there're a couple of issues with this.
1. There is a padding in the dockspace, which I don't know how to get rid of.
2. I don't want those title bars to be there, if I add the `ImGuiWindowFlags_NoTitleBar` flag to the two panels inside (which are just windows), the title bar disappears but only if the window is not docked, once you dock it, the title bar shows up again. And I also tried add the `ImGuiDockNodeFlags_AutoHideTabBar` but "hiding the tab bar" does not solve the problem, because there is still the lil triangle where you can click and it show up again, I don't want that, I want the title bar to be completely gone, basically just a V splitter inside the window, is that possible?
3. I would also need the two panels to be not un-dockable, if that's a thing. There is `ImGuiDockNodeFlags_KeepAliveOnly` which has this comment "Don't display the dockspace node but keep it alive. Windows docked into this dockspace node won't be undocked", but when I add this, I don't even see the panels, and I don't see the splitter neither.

Basically, what I'm looking for would be something like this:

![image](https://user-images.githubusercontent.com/80007185/207665394-8817f520-7ce5-4ad2-94fa-52c5b0b0dae0.png)
Which I created with `SplitterBehaviour()`, but as I said, that seems to be only visual, it don't work when you start adding components in.

Btw, is this accomplished with columns or something like that?

Any help is appreciated, thx :)

**Comment by marpe** [comment]
`SplitterBehaviour()` is not only visual, but imho docking is cleaner. Haven't been able to test this code out but it should give you an idea for how to set it up.

```c++
bool const shouldDrawWindowContents = ImGui::Begin("MyWindow");
ImGuiID const dockspaceID = ImGui::GetID("MyWindow_DockSpace");
ImGuiWindowClass workspaceWindowClass;
workspaceWindowClass.ClassId = dockspaceID;
workspaceWindowClass.DockingAllowUnclassed = false;

if (!ImGui::DockBuilderGetNode(dockspaceID)) {
    ImGui::DockBuilderAddNode(dockspaceID, ImGuiDockNodeFlags_DockSpace | ImGuiDockNodeFlags_NoWindowMenuButton |
                                           ImGuiDockNodeFlags_NoCloseButton);
    ImGui::DockBuilderSetNodeSize(dockspaceID, ImGui::GetContentRegionAvail());

    //

    ImGuiID leftDockID = 0, rightDockID = 0;
    ImGui::DockBuilderSplitNode(dockspaceID, ImGuiDir_Left, 0.5f, &leftDockID, &rightDockID);

    ImGuiDockNode* pLeftNode = ImGui::DockBuilderGetNode( leftDockID );
    ImGuiDockNode* pRightNode = ImGui::DockBuilderGetNode( rightDockID );
    pLeftNode->LocalFlags |= ImGuiDockNodeFlags_NoTabBar | ImGuiDockNodeFlags_NoDockingSplitMe | ImGuiDockNodeFlags_NoDockingOverMe;
    pRightNode->LocalFlags |= ImGuiDockNodeFlags_NoTabBar | ImGuiDockNodeFlags_NoDockingSplitMe | ImGuiDockNodeFlags_NoDockingOverMe;

    // Dock windows
    ImGui::DockBuilderDockWindow("LeftWindow", leftDockID);
    ImGui::DockBuilderDockWindow("RightWindow", rightDockID);

    //

    ImGui::DockBuilderFinish(dockspaceID);
}

ImGuiDockNodeFlags const dockFlags = shouldDrawWindowContents ? ImGuiDockNodeFlags_None
                                                              : ImGuiDockNodeFlags_KeepAliveOnly;
ImGui::DockSpace(dockspaceID, ImGui::GetContentRegionAvail(), dockFlags, &workspaceWindowClass);

ImGui::End();

if (ImGui.Begin("LeftWindow"))
{
    //
}

ImGui.End();

if (ImGui.Begin("RightWindow"))
{
    //
}

ImGui.End();
```

**Comment by carlosdr02** [comment]
@marpe It works perfectly, but is it just me or the API for doing this kind of things is quite immature? I mean, I've never heard of `ImGuiWindowClass` nor `ImGui::DockBuilder...()`, not to mention that all of this is in `<imgui_internal.h>`.

@ocornut Is this the standard way of doing this right now if you don't mind me asking?

**Comment by developer239** [comment]
@carlosdr02 Because the example is incomplete and it didn't even mention that the API only lives in `docking` branch.

It took me a while to make it work but here is a minimal working example with SDL and SDL_renderer.

It uses [dock builder example snippet](https://gist.github.com/PossiblyAShrub/0aea9511b84c34e191eaa90dd7225969) you can find it in my own repository with CMake that should make it easy to run here: [cmake-sdl2-imgui](https://github.com/developer239/cmake-sdl2-imgui/tree/docking)

```
#include <SDL.h>

#include "externals/imgui/backends/imgui_impl_sdl.h"
#include "externals/imgui/backends/imgui_impl_sdlrenderer.h"
#include "externals/imgui/imgui.h"
#include "externals/imgui/imgui_internal.h"

int main() {
  SDL_Window* window;

  SDL_Init(SDL_INIT_EVERYTHING);

  SDL_DisplayMode displayMode;
  SDL_GetCurrentDisplayMode(0, &displayMode);

  window = SDL_CreateWindow(
      nullptr,
      SDL_WINDOWPOS_CENTERED,
      SDL_WINDOWPOS_CENTERED,
      800,
      600,
      0
  );

  SDL_Renderer* renderer = SDL_CreateRenderer(window, -1, 0);

  ImGui::CreateContext();
  ImGuiIO& io = ImGui::GetIO();

  io.ConfigFlags |= ImGuiConfigFlags_DockingEnable;

  ImGui_ImplSDL2_InitForSDLRenderer(window, renderer);
  ImGui_ImplSDLRenderer_Init(renderer);

  bool shouldQuit = false;
  SDL_Event event;

  while (!shouldQuit) {
    while (SDL_PollEvent(&event)) {
      ImGui_ImplSDL2_ProcessEvent(&event);

      if (event.type == SDL_QUIT) {
        shouldQuit = true;
      }
    }

    ImGui_ImplSDLRenderer_NewFrame();
    ImGui_ImplSDL2_NewFrame();

    ImGui::NewFrame();
    static ImGuiDockNodeFlags dockspace_flags =
        ImGuiDockNodeFlags_PassthruCentralNode;

    // We are using the ImGuiWindowFlags_NoDocking flag to make the parent
    // window not dockable into, because it would be confusing to have two
    // docking targets within each others.
    ImGuiWindowFlags window_flags =
        ImGuiWindowFlags_MenuBar | ImGuiWindowFlags_NoDocking;

    ImGuiViewport* viewport = ImGui::GetMainViewport();
    ImGui::SetNextWindowPos(viewport->Pos);
    ImGui::SetNextWindowSize(viewport->Size);
    ImGui::SetNextWindowViewport(viewport->ID);
    ImGui::PushStyleVar(ImGuiStyleVar_WindowRounding, 0.0f);
    ImGui::PushStyleVar(ImGuiStyleVar_WindowBorderSize, 0.0f);
    window_flags |= ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoCollapse |
                    ImGuiWindowFlags_NoResize | ImGuiWindowFlags_NoMove;
    window_flags |=
        ImGuiWindowFlags_NoBringToFrontOnFocus | ImGuiWindowFlags_NoNavFocus;

    // When using ImGuiDockNodeFlags_PassthruCentralNode, DockSpace() will
    // render our background and handle the pass-thru hole, so we ask Begin() to
    // not render a background.
    if (dockspace_flags & ImGuiDockNodeFlags_PassthruCentralNode)
      window_flags |= ImGuiWindowFlags_NoBackground;

    // Important: note that we proceed even if Begin() returns false (aka window
    // is collapsed). This is because we want to keep our DockSpace() active. If
    // a DockSpace() is inactive, all active windows docked into it will lose
    // their parent and become undocked. We cannot preserve the docking
    // relationship between an active window and an inactive docking, otherwise
    // any change of dockspace/settings would lead to windows being stuck in
    // limbo and never being visible.
    ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, ImVec2(0.0f, 0.0f));
    ImGui::Begin("DockSpace", nullptr, window_flags);
    ImGui::PopStyleVar();
    ImGui::PopStyleVar(2);

    // DockSpace
    ImGuiIO& io = ImGui::GetIO();
    if (io.ConfigFlags & ImGuiConfigFlags_DockingEnable) {
      ImGuiID dockspace_id = ImGui::GetID("MyDockSpace");
      ImGui::DockSpace(dockspace_id, ImVec2(0.0f, 0.0f), dockspace_flags);

      static auto first_time = true;
      if (first_time) {
        first_time = false;

        ImGui::DockBuilderRemoveNode(dockspace_id
        );  // clear any previous layout
        ImGui::DockBuilderAddNode(
            dockspace_id,
            dockspace_flags | ImGuiDockNodeFlags_DockSpace
        );
        ImGui::DockBuilderSetNodeSize(dockspace_id, viewport->Size);

        // split the dockspace into 2 nodes -- DockBuilderSplitNode takes in the
        // following args in the following order
        //   window ID to split, direction, fraction (between 0 and 1), the
        //   final two setting let's us choose which id we want (which ever one
        //   we DON'T set as NULL, will be returned by the function)
        //                                                              out_id_at_dir
        //                                                              is the
        //                                                              id of
        //                                                              the node
        //                                                              in the
        //                                                              direction
        //                                                              we
        //                                                              specified
        //                                                              earlier,
        //                                                              out_id_at_opposite_dir
        //                                                              is in
        //                                                              the
        //                                                              opposite
        //                                                              direction
        auto dock_id_left = ImGui::DockBuilderSplitNode(
            dockspace_id,
            ImGuiDir_Left,
            0.2f,
            nullptr,
            &dockspace_id
        );
        auto dock_id_down = ImGui::DockBuilderSplitNode(
            dockspace_id,
            ImGuiDir_Down,
            0.25f,
            nullptr,
            &dockspace_id
        );

        // we now dock our windows into the docking node we made above
        ImGui::DockBuilderDockWindow("Down", dock_id_down);
        ImGui::DockBuilderDockWindow("Left", dock_id_left);
        ImGui::DockBuilderFinish(dockspace_id);
      }
    }

    ImGui::End();

    ImGui::Begin("Left");
    ImGui::Text("Hello, left!");
    ImGui::End();

    ImGui::Begin("Down");
    ImGui::Text("Hello, down!");
    ImGui::End();

    ImGui::Render();

    SDL_SetRenderDrawColor(renderer, 0, 0, 100, 0);
    SDL_RenderClear(renderer);

    ImGui_ImplSDLRenderer_RenderDrawData(ImGui::GetDrawData());
    SDL_RenderPresent(renderer);
  }

  ImGui_ImplSDLRenderer_Shutdown();
  ImGui_ImplSDL2_Shutdown();
  ImGui::DestroyContext();

  SDL_DestroyRenderer(renderer);
  SDL_DestroyWindow(window);
  SDL_Quit();
}
```
